src/mlpack/methods/cf/normalization/combined_normalization.hpp

```diff
@@ -0,0 +1,231 @@
+/**
+ * @file combined_normalization.hpp
+ * @author Wenhao Huang
+ *
+ * CombinedNormalization is a class template for performing a sequence of data
+ * normalization methods which are specified by template parameter.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_CF_NORMALIZATION_COMBINED_NORMALIZATION_HPP
+#define MLPACK_METHODS_CF_NORMALIZATION_COMBINED_NORMALIZATION_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace cf {
+
+/**
+ * This normalization class performs a sequence of normalization methods on
+ * raw ratings.
+ *
+ * An example of how to use CombinedNormalization in CF is shown below:
+ *
+ * @code
+ * extern arma::mat data; // (user, item, rating) table
+ * extern arma::Col<size_t> users; // users seeking recommendations
+ * arma::Mat<size_t> recommendations; // Recommendations
+ *
+ * CFType<CombinedNormalization<
+ *            OverallMeanNormalization,
+ *            UserMeanNormalization,
+ *            ItemMeanNormalization>> cf(data);
+ *
+ * // Generate 10 recommendations for all users.
+ * cf.GetRecommendations(10, recommendations);
+ * @endcode
+ */
+template<typename... NormalizationTypes>
+class CombinedNormalization
+{
+ public:
+  using TupleType = std::tuple<NormalizationTypes...>;
+
+  // Empty constructor.
+  CombinedNormalization() { }
+
+  /**
+   * Normalize the data by calling Normalize() in each normalization object.
+   *
+   * @param data Input dataset in the form of coordinate list.
+   */
+  void Normalize(arma::mat& data)
+  {
+    SequenceNormalize<0>(data);
+  }
+
+  /**
+   * Normalize the data by calling Normalize() in each normalization object.
+   *
+   * @param cleanedData Sparse matrix data.
+   */
+  void Normalize(arma::sp_mat& cleanedData)
+  {
+    SequenceNormalize<0>(cleanedData);
```

Comment:
Haven't really thought about it but it might be worth to take a look into ``boost::fusion`` ``for_each``.

Reply:
I think it could work but `boost::fusion` isn't one of mlpack's dependencies yet.